### PR TITLE
Rule S4426: Add support for target typed new expressions

### PR DESCRIPTION
**Pull request by dmitry-golovinov-sonarsource**

### CODE REVIEW

**Comment by dmitry-golovinov-sonarsource** [comment]
Only the last commit is to be reviewed

File: analyzers/src/SonarAnalyzer.CSharp/Rules/CryptographicKeyShouldNotBeTooShort.cs [hunk 1]
```diff
@@ -98,17 +100,17 @@ protected override void Initialize(SonarAnalysisContext context)
             context.RegisterSyntaxNodeActionInNonGenerated(
                 c =>
                 {
-                    var objectCreation = (ObjectCreationExpressionSyntax)c.Node;
-                    if (!(c.SemanticModel.GetTypeInfo(objectCreation).Type is ITypeSymbol containingType))
+                    var objectCreation = ObjectCreationFactory.Create(c.Node);
+                    if (!(c.SemanticModel.GetTypeInfo(objectCreation.Expression).Type is ITypeSymbol containingType))
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/CryptographicKeyShouldNotBeTooShort.cs:104]: Let's refactor this old code to positive condition with actions in if block.
The `!condition` with `containingType` sideeffect and `return` is a double indirection (hard to read)

Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/CryptographicKeyShouldNotBeTooShort.cs:104]: Can the condition be covered with nonexistent type?
```
var c = new UndefinedType(); // Error [CS0264]
```

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
Minor comments

File: analyzers/src/SonarAnalyzer.CSharp/Rules/CryptographicKeyShouldNotBeTooShort.cs (hunk 1 above)
Comment (dmitry-golovinov-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/CryptographicKeyShouldNotBeTooShort.cs]: Unfortunately, didn't help

**Review by pavel-mikula-sonarsource — APPROVED** [review]
LGTM

File: analyzers/src/SonarAnalyzer.CSharp/Rules/CryptographicKeyShouldNotBeTooShort.cs [hunk 2]
```diff
@@ -98,17 +100,15 @@ protected override void Initialize(SonarAnalysisContext context)
             context.RegisterSyntaxNodeActionInNonGenerated(
                 c =>
                 {
-                    var objectCreation = (ObjectCreationExpressionSyntax)c.Node;
-                    if (!(c.SemanticModel.GetTypeInfo(objectCreation).Type is ITypeSymbol containingType))
+                    var objectCreation = ObjectCreationFactory.Create(c.Node);
+                    if (c.SemanticModel.GetTypeInfo(objectCreation.Expression).Type is { } containingType)
```
[the remark this answers is not in the document]
Comment (dmitry-golovinov-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/CryptographicKeyShouldNotBeTooShort.cs]: Seems like it.

[the remark this answers is not in the document]
Comment (dmitry-golovinov-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/CryptographicKeyShouldNotBeTooShort.cs]: Indeed, that one returned empty on incorrect syntax` semanticModel.GetTypeInfo(objectCreation).Type.Name`

**Pull request merged**